src/common/cockpitwebserver.c

```diff
@@ -1028,46 +1028,60 @@ on_request_input (GObject *pollable_input,
   gsize length;
   gssize count;
 
-  length = request->buffer->len;
-  g_byte_array_set_size (request->buffer, length + 4096);
-
-  count = g_pollable_input_stream_read_nonblocking (input, request->buffer->data + length,
-                                                    4096, NULL, &error);
-  if (count < 0)
+  /* read all pending data */
+  for (;;)
     {
-      g_byte_array_set_size (request->buffer, length);
+      length = request->buffer->len;
+      g_byte_array_set_size (request->buffer, length + 4096);
 
-      /* Just wait and try again */
-      if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_WOULD_BLOCK))
+      count = g_pollable_input_stream_read_nonblocking (input, request->buffer->data + length,
+                                                        4096, NULL, &error);
+      if (count < 0)
         {
+          g_byte_array_set_size (request->buffer, length);
+
+          if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_WOULD_BLOCK))
+            {
+              /* finished reading everything that was pending */
+              g_error_free (error);
+              break;
+            }
+
+          if (!should_suppress_request_error (error, length))
+            g_message ("couldn't read from connection: %s", error->message);
+
+          cockpit_request_finish (request);
           g_error_free (error);
-          return TRUE;
+          return FALSE;
         }
 
-      if (!should_suppress_request_error (error, length))
-        g_message ("couldn't read from connection: %s", error->message);
+      g_byte_array_set_size (request->buffer, length + count);
 
-      cockpit_request_finish (request);
-      g_error_free (error);
-      return FALSE;
-    }
-
-  g_byte_array_set_size (request->buffer, length + count);
+      if (count == 0)
+        {
+          if (request->eof_okay)
+            close_io_stream (request->io);
+          else
+            g_debug ("caller closed connection early");
+          if (request->buffer->len == 0)
+            {
+              cockpit_request_finish (request);
+              return FALSE;
+            }
+          else
+            {
+              break;
+            }
+        }
 
-  if (count == 0)
-    {
-      if (request->eof_okay)
-        close_io_stream (request->io);
-      else
-        g_debug ("caller closed connection early");
-      cockpit_request_finish (request);
-      return FALSE;
+      /* Once we receive data EOF is unexpected (until possible next request) */
+      request->eof_okay = FALSE;
     }
 
-  /* Once we receive data EOF is unexpected (until possible next request) */
-  request->eof_okay = FALSE;
-
-  return parse_and_process_request (request);
+  if (request->buffer->len > 0)
+    return parse_and_process_request (request);
```

Comment:
@martinpitt I think that this should be called within the loop for each successful read. It will both simplify the code, and also remain true to the original intent of how that function affects the request structure.